### PR TITLE
[model] feat: [transformers-v5] Add ViT support for Qwen3.5

### PR DESCRIPTION
**Pull request by deerlu**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

This PR adds Vision Transformer (ViT) support for Qwen3.5 models with transformers v5. It includes:
- Qwen3.5 model patches with vision encoder support
- New data processing function process_sample_qwen3_vl_transformers_v5 for Qwen3.5 VL models
- Remove the manually added `system_prompt` in Qwen3VLChatTemplate to align with the HF implementation
- Test cases for validating Qwen3.5 VL data processing
- Training config for Qwen3.5 VL model

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `perf`, `ops`, `parallel`
  - If this PR involves multiple modules, separate them with `,` like `[ci, data, model]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

- Added unit test test_qwen3_5_processor_comparison in tests/data/multimodal/test_vlm_data_process.py
  - Test validates that HF and VeOmni processors produce identical outputs (input_ids, attention_mask, pixel_values, position_ids, mm_token_type_ids, etc.)
- Run test_models_patch with transformers v5 to validate Qwen3.5 forward/backward parity
- Run test_models_patch with transformers v4 to confirm existing models are unaffected
- End-to-end SFT training on 4xGPU with configs/multimodal/qwen3_5/qwen3_5_vl.yaml
  - Successfully completed 2 epochs (40 steps) with loss
  decreasing
image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```bash
bash train.sh tasks/train_vlm.py configs/multimodal/qwen3_5/qwen3_5_vl.yaml \
    --model.model_path ./Qwen/Qwen3.5-0.8B
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by deerlu** [comment]
Follow-up: #523

File: veomni/data/data_transform.py [hunk 1]
```diff
@@ -337,6 +337,88 @@ def process_sample_qwen3_vl(
...
+
+    tokenized_example = {
+        k: (v if isinstance(v, torch.Tensor) else torch.tensor(v)) for k, v in tokenized_example.items()
+    }
+
+    input_ids = tokenized_example["input_ids"]
+    tokenized_example["image_mask"] = input_ids == IMAGE_INPUT_INDEX
+    tokenized_example["video_mask"] = input_ids == VIDEO_INPUT_INDEX
+    mm_token_type_ids = torch.zeros_like(input_ids)
+    mm_token_type_ids[tokenized_example["image_mask"]] = 1
```
Comment (Coach257) [inline veomni/data/data_transform.py]: Mv `mm_token_type_ids` to `process_sample_qwen3vl` & `process_sample_qwen2vl` and merge these two process functions. Qwenvl series in Transformers v5.3 all use `mm_token_type_ids` to build position_ids. https://github.com/huggingface/transformers/blob/aad13b87ed59f2afcfaebc985f403301887a35fc/src/transformers/models/qwen3_vl/modeling_qwen3_vl.py#L1040
https://github.com/huggingface/transformers/blob/aad13b87ed59f2afcfaebc985f403301887a35fc/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py#L1115

File: tests/e2e/test_e2e_parallel.py [hunk 2]
```diff
@@ -183,8 +172,20 @@ def main(
...
     ),
+    pytest.param(
+        "qwen3_5",
+        "./tests/toy_config/qwen3_5_toy/config.json",
+        False,  # is_moe
+        _DEFAULT_RTOL,
+        _DEFAULT_ATOL,
+        # TODO: remove max_sp_size limit once we support SP in qwen3_5.
+        # In addition, since SP is limited, there is only one test case now so nothing is compared.
+        1,  # max_sp_size
```
Comment (piyifan123) [inline tests/e2e/test_e2e_parallel.py]: i think SP is already supported. we can enable this now or a follow up MR.

File: veomni/data/data_transform.py (hunk 1 above)
Comment (deerlu) [reply veomni/data/data_transform.py]: done

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 3]
```diff
@@ -1296,67 +1326,62 @@ def rot_pos_emb(self, grid_thw: torch.Tensor) -> torch.Tensor:
...
-            pos_embed = (
-                pos_embed.view(t, h // merge_size, merge_size, w // merge_size, merge_size, -1)
-                .permute(0, 1, 3, 2, 4, 5)
-                .flatten(0, 4)
-            )
-            patch_pos_embeds_permute.append(pos_embed)
-        patch_pos_embeds = torch.cat(patch_pos_embeds_permute)
-        return patch_pos_embeds
+        """
+        Copied and adapted from https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/qwen3_vl.py#L474
```
Comment (TimYangst) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: nit: use the url with the commit id to link to the orignal code snapshot? as the `main` branch version may be changed.

File: docs/examples/qwen3_5.md [hunk 4]
```diff
@@ -42,6 +65,17 @@ python3 scripts/download_hf_model.py \
 
 ## Start training on GPU
 
+### Qwen3.5-27B VL Training
```
Comment (piyifan123) [inline docs/examples/qwen3_5.md]: nit: use https://huggingface.co/Qwen/Qwen3.5-9B as an example since 27B is too big to fit in one 8x80GB machine.

File: tests/data/multimodal/test_vlm_data_process.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        raise ValueError(
+            f"Unknown process_sample function: {config.process_sample_func_name}. "
+            f"Available functions: {list(PROCESS_SAMPLE_FUNCTION_MAP.keys())}"
+        )
+
+    # Prepare sample in VeOmni format
+    sample = {
+        "source_name": "sharegpt4v_sft",
+        "conversations": [
+            {"from": "human", "value": f"<image>{config.test_text}"},
```
Comment (piyifan123) [inline tests/data/multimodal/test_vlm_data_process.py:163]: nit: can we add more messages like user, and assistant to be more realistic?

File: tests/data/multimodal/test_vlm_data_process.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    )
+    output = outputs[0]
+
+    # Move tensors to device
+    output = {k: (v.to(device) if isinstance(v, torch.Tensor) else v) for k, v in output.items()}
+
+    return output
+
+
+@pytest.mark.parametrize(
```
Comment (piyifan123) [inline tests/data/multimodal/test_vlm_data_process.py:191]: thanks for adding the test to verify processor logic!

File: veomni/data/data_transform.py [hunk 7]
```diff
@@ -260,81 +287,43 @@ def process_sample_qwen2_5_vl(
     return [tokenized_example]
 
 
+@DATA_TRANSFORM_REGISTRY.register("qwen2_vl")
+@DATA_TRANSFORM_REGISTRY.register("qwen2_5_vl")
 @DATA_TRANSFORM_REGISTRY.register("qwen3_vl")
 @DATA_TRANSFORM_REGISTRY.register("qwen3_vl_moe")
-def process_sample_qwen3_vl(
+def process_sample_qwen_vl_transformer_v4(
```
Comment (piyifan123) [inline veomni/data/data_transform.py]: if we train those models with transformers v5, would it cause an issue? should the register (or the returned function) based on the transformers version in the env?

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py (hunk 3 above)
Comment (piyifan123) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: nit: add more details in comment on why we change this to a vllm qwen3vl implementation instead of using transformers impl.

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 8]
```diff
@@ -1374,15 +1399,14 @@ def forward(self, hidden_states: torch.Tensor, grid_thw: torch.Tensor, **kwargs)
         hidden_states = self.patch_embed(hidden_states)
 
         pos_embeds = self.fast_pos_embed_interpolate(grid_thw)
-        hidden_states = hidden_states + pos_embeds
 
-        rotary_pos_emb = self.rot_pos_emb(grid_thw)
+        # --- Patch.1 ---
+        if get_parallel_state().sp_enabled:
+            # We need to do padding here because of hidden_states did padding with pad_scale=4
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: nit: add a comment to explain "magic number" 4.

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 9]
```diff
@@ -1671,67 +1744,25 @@ def get_video_features(
...
-            )
-
-        n_video_tokens = special_video_mask.sum()
-        special_video_mask = special_video_mask.unsqueeze(-1).expand_as(inputs_embeds).to(inputs_embeds.device)
-        if video_features is not None:
-            torch_compilable_check(
-                inputs_embeds[special_video_mask].numel() == video_features.numel(),
-                f"Video features and video tokens do not match, tokens: {n_video_tokens}, features: {video_features.shape[0]}",
-            )
+        special_image_mask = input_ids == self.config.image_token_id
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py:1792]: nit: add patched comments to explain the change here?

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 10]
```diff
@@ -1671,67 +1744,25 @@ def get_video_features(
...
         pixel_values = pixel_values.type(self.visual.dtype)
-        vision_output: BaseModelOutputWithPooling = self.visual(
-            pixel_values, grid_thw=image_grid_thw, return_dict=True, **kwargs
-        )
-        image_embeds = vision_output.pooler_output
-        split_sizes = (image_grid_thw.prod(-1) // self.visual.spatial_merge_size**2).tolist()
-        image_embeds = torch.split(image_embeds, split_sizes)
-        vision_output.pooler_output = image_embeds
-
+        vision_output: BaseModelOutputWithPooling = self.visual(pixel_values, grid_thw=image_grid_thw, return_dict=True)
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: nit: add patched comments to explain the change here?

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 11]
```diff
@@ -1829,7 +1966,17 @@ def forward(
                 inputs_embeds=inputs_embeds,
                 attention_mask=attention_mask,
                 past_key_values=past_key_values,
+                mm_token_type_ids=mm_token_type_ids,
             )
+        else:
+            # --- Patch.3 ---
+            if position_ids.dim() == 3 and position_ids.shape[1] == 3:
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py:2009]: what's this for? @Coach257 BTW, what position ids do we use for qwenvl model family, I remember we changed from transformers setting before?

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py [hunk 12]
```diff
@@ -1374,15 +1399,14 @@ def forward(self, hidden_states: torch.Tensor, grid_thw: torch.Tensor, **kwargs)
         hidden_states = self.patch_embed(hidden_states)
 
         pos_embeds = self.fast_pos_embed_interpolate(grid_thw)
-        hidden_states = hidden_states + pos_embeds
 
-        rotary_pos_emb = self.rot_pos_emb(grid_thw)
+        # --- Patch.1 ---
```
Comment (TimYangst) [inline veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: Could you revisit the Patch comments and add a brief summary at the top of the function explaining what each Patch.#N is addressing?

Here is an example of how this can be organized:
https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py#L57-L63

I think an AI coding tool could help quickly restructure these comments and keep them consistent.

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py (hunk 11 above)
Comment (Coach257) [reply veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py:2009]: As we precompute `position_ids` and pack it, the shape is `[bs, dim, seqlen]`. Qwenvl receives `[dim, bs, seqlen]` so we always need to transpose here.

**Comment by Coach257** [comment]
LGTM

File: tests/data/multimodal/test_vlm_data_process.py (hunk 5 above)
Comment (deerlu) [reply tests/data/multimodal/test_vlm_data_process.py:163]: "from": "human" is mapped to the "user" role.  I tried to add a assistant role to the test, but this exposed some inconsistencies with HF's <think> block processing, will address this issue in a separate follow-up PR

File: veomni/data/data_transform.py (hunk 7 above)
Comment (deerlu) [reply veomni/data/data_transform.py]: Done. Data transform methods are now registered according to the Transformers version in the env

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py (hunk 8 above)
Comment (deerlu) [reply veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: done

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py (hunk 10 above)
Comment (deerlu) [reply veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: done

File: veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py (hunk 3 above)
Comment (deerlu) [reply veomni/models/transformers/qwen3_5/generated/patched_modeling_qwen3_5_gpu.py]: done
